### PR TITLE
[FSTORE-1118] Pandas arrow type dataframes cannot be inserted into feature group

### PR DESCRIPTION
**Pull request by manu-sj**
This PR fixes issue of pyarrow types introduced as part of pandas 2.0 not being insterted into the feature store.

- Changes were made in this PR adds support to convert for arrow datatype introduced in pandas 2.0.
- This PR changes the way the conversion between pandas datatypes were performed to increase readability and maintainability of the code. Dicussion done in design log - https://hopsworks.atlassian.net/wiki/spaces/FST/pages/279314450/Conversion+of+Pandas+arrow+datatypes+to+hopsworks+internal+Datatypes
- Testing involved checking if all types and corresponding arrow types mentioned as part of hopsworks documentation is inserted into the feature group.
- Some datatypes present introducted as part of pyarrow in pandas are not added since they were not explicitly mentioned in the hopsworks documentation. The datatypes not added are mentioned are - uint64, large_binary, large_string, large_utf8, large_list, fixed_size_list (pyarrow list with max size mentioned), maps, dictonary (used for efficent array storage in pyarrow)  

JIRA Issue: - https://hopsworks.atlassian.net/browse/FSTORE-1118

Priority for Review: -

Related PRs: -

**How Has This Been Tested?**

- [x] Unit Tests
- [ ] Integration Tests
- [x] Manual Tests on VM

**Checklist For The Assigned Reviewer:**

```
- [ ] Checked if merge conflicts with master exist
- [ ] Checked if stylechecks for Java and Python pass
- [ ] Checked if all docstrings were added and/or updated appropriately
- [ ] Ran spellcheck on docstring
- [ ] Checked if guides & concepts need to be updated
- [ ] Checked if naming conventions for parameters and variables were followed
- [ ] Checked if private methods are properly declared and used
- [ ] Checked if hard-to-understand areas of code are commented
- [ ] Checked if tests are effective
- [ ] Built and deployed changes on dev VM and tested manually
- [x] (Checked if all type annotations were added and/or updated appropriately)
```

### CODE REVIEW

File: python/hsfs/engine/python.py [hunk 1]
```diff
@@ -468,11 +469,11 @@ def convert_to_default_dataframe(self, dataframe):
     def parse_schema_feature_group(self, dataframe, time_travel_format=None):
         arrow_schema = pa.Schema.from_pandas(dataframe)
         features = []
-        for feat_name, feat_type in dataframe.dtypes.items():
+        for feat_name, _ in dataframe.dtypes.items():
```
Comment (bubriks) [inline python/hsfs/engine/python.py:472]:
```suggestion
        for feat_name in arrow_schema.names:
```

**Review by bubriks — CHANGES_REQUESTED** [review]

File: python/tests/engine/test_python.py [hunk 2]
```diff
@@ -1102,7 +1105,7 @@ def test_convert_simple_pandas_type_uint8(self):
 
         # Act
         result = python_engine._convert_simple_pandas_dtype_to_offline_type(
-            dtype=np.dtype("uint8")
+            arrow_type=pa.from_numpy_dtype(np.dtype("uint8"))
```
Comment (bubriks) [inline python/tests/engine/test_python.py:1108]: Not a big deal but instead of doing `pa.from_numpy_dtype(np.dtype("uint8"))` it would be cleaner to do `pa.uint8()` (this idea applies in multiple places in `test_python.py`)

File: python/tests/engine/test_python.py [hunk 3]
```diff
@@ -1567,6 +1572,346 @@ def test_infer_type_pyarrow_struct_with_list_of_struct_fields(self):
...
+        pdf = pd.DataFrame({"data": pd.Series([1, 5], dtype="int32[pyarrow]")})
+        arrow_schema = pa.Schema.from_pandas(pdf)
+
+        python_engine = python.Engine()
+
+        # Act
+        arrow_type = python_engine._convert_simple_pandas_dtype_to_offline_type(
+            arrow_schema.field("data").type
+        )
+        print(arrow_type)
```
Comment (bubriks) [inline python/tests/engine/test_python.py:1618]: `print` should be removed (multiple places)

File: python/tests/engine/test_python.py [hunk 4]
```diff
@@ -1567,6 +1572,346 @@ def test_infer_type_pyarrow_struct_with_list_of_struct_fields(self):
...
+        # Arrange
+        pdf = pd.DataFrame({"data": pd.Series([1, 5], dtype="int32[pyarrow]")})
+        arrow_schema = pa.Schema.from_pandas(pdf)
+
+        python_engine = python.Engine()
+
+        # Act
+        arrow_type = python_engine._convert_simple_pandas_dtype_to_offline_type(
+            arrow_schema.field("data").type
+        )
```
Comment (bubriks) [inline python/tests/engine/test_python.py:1617]:
```suggestion
        python_engine = python.Engine()

        # Act
        arrow_type = python_engine._convert_simple_pandas_dtype_to_offline_type(
            pa.int32()
        )
```

Comment (bubriks) [reply python/tests/engine/test_python.py:1617]: could be done like this in multiple tests, instead of creating data frame.

File: python/tests/engine/test_python.py [hunk 5]
```diff
@@ -1567,6 +1572,346 @@ def test_infer_type_pyarrow_struct_with_list_of_struct_fields(self):
...
+        print(arrow_type)
+
+        # Assert
+        assert arrow_type == "int"
+
+    def test_infer_type_pandas_pyarrow_int32(self):
```
Comment (bubriks) [inline python/tests/engine/test_python.py:1607]: it looks like these tests are duplicates, for example we already have this case covered: https://github.com/logicalclocks/feature-store-api/pull/1175/files#diff-01253ce273d2953406cf0ee612bf433ed4fdaae72f49126492cfe23841beb082L1147

Make sure there are no duplicates. And our naming for test methods is `test_[testedMethodName]_[condition]`. In this case it should be something like`test_convert_simple_pandas_dtype_to_offline_type_int32`

**Review by bubriks — CHANGES_REQUESTED** [review]

File: python/tests/engine/test_python.py [hunk 6]
```diff
@@ -1088,12 +1088,13 @@ def test_convert_pandas_type_object(self, mocker):
         python_engine = python.Engine()
 
         # Act
+        python_engine._convert_pandas_dtype_to_offline_type(arrow_type=pa.struct({}))
         python_engine._convert_pandas_dtype_to_offline_type(
-            dtype=np.dtype("O"), arrow_type=None
+            arrow_type=pa.list_(pa.int32())
         )
```
Comment (bubriks) [inline python/tests/engine/test_python.py:1094]: should be removed.

File: python/tests/engine/test_python.py [hunk 7]
```diff
@@ -1088,12 +1088,13 @@ def test_convert_pandas_type_object(self, mocker):
...
 
         # Assert
-        assert mock_python_engine_infer_type_pyarrow.call_count == 1
+        assert mock_python_engine_infer_type_pyarrow.call_count == 2
```
Comment (bubriks) [inline python/tests/engine/test_python.py:1097]:
```suggestion
        assert mock_python_engine_infer_type_pyarrow.call_count == 1
```

File: python/tests/engine/test_python.py [hunk 8]
```diff
@@ -1088,12 +1088,13 @@ def test_convert_pandas_type_object(self, mocker):
         python_engine = python.Engine()
 
         # Act
+        python_engine._convert_pandas_dtype_to_offline_type(arrow_type=pa.struct({}))
```
Comment (bubriks) [inline python/tests/engine/test_python.py:1091]: we should check the results of convert for `struct` and `list` types, the same way as for `int64` for example.

Comment (bubriks) [reply python/tests/engine/test_python.py:1091]: the patching can be removed, the call count is not as important as the result. The same applies in other tests as well, i think it was a mistake that we patched as much as we did so it should be fixed.

File: python/tests/engine/test_python.py (hunk 6 above)
Comment (bubriks) [reply python/tests/engine/test_python.py:1094]: this test is for `struct` not `list`

File: python/tests/engine/test_python.py (hunk 8 above)
Comment (bubriks) [reply python/tests/engine/test_python.py:1091]: so for example for list it would look like this:
```
    def test_convert_pandas_object_type_to_offline_type_list(self, mocker):
        # Arrange
        python_engine = python.Engine()

        # Act
        result = python_engine._convert_pandas_object_type_to_offline_type(
            arrow_type=pa.list_(pa.int8())
        )

        # Assert
        assert result == "array<int>"
```

**Review by bubriks — CHANGES_REQUESTED** [review]

File: python/tests/engine/test_python.py (hunk 8 above)
Comment (manu-sj) [reply python/tests/engine/test_python.py]: I removed the entire function `test_convert_pandas_type_object` since there was already another function - `test_infer_type_pyarrow_struct` that does the test for an empty sturctures. Similarly `test_convert_pandas_type` was also removed since there was another function `test_convert_simple_pandas_type_int64` that does the same test when the call count is removed.

File: python/tests/engine/test_python.py [hunk 9]
```diff
@@ -1258,39 +1236,30 @@ def test_convert_simple_pandas_type_other(self):
...
         # Act
         result = python_engine._convert_pandas_object_type_to_offline_type(
             arrow_type=pa.list_(pa.int8())
         )
 
         # Assert
-        assert result == "array<test>"
+        assert result == "array<int>"
 
     def test_infer_type_pyarrow_struct(self, mocker):
```
Comment (bubriks) [inline python/tests/engine/test_python.py:1259]: one minor improvement for this test:
```
    def test_infer_type_pyarrow_struct(self):
        # Arrange
        python_engine = python.Engine()

        # Act
        result = python_engine._convert_pandas_object_type_to_offline_type(
            arrow_type=pa.struct([pa.field('f1', pa.int32())])
        )

        # Assert
        assert result == "struct<f1:int>"
```

Comment (bubriks) [reply python/tests/engine/test_python.py:1259]: also optionally you can remove `mocker` args from test methods where they are not needed. (but ok if you wish to not do it :) )

File: python/hsfs/engine/python.py [hunk 10]
```diff
@@ -82,6 +83,31 @@
...
+    **dict.fromkeys([pa.bool_()], "boolean"),
+    **dict.fromkeys(
+        [
+            pa.string(),
+            pa.dictionary(value_type=pa.string(), index_type=pa.int8(), ordered=True),
+            pa.dictionary(value_type=pa.string(), index_type=pa.int8(), ordered=False),
+        ],
+        "string",
+    ),  # Category type in pandas stored as dictinoary in pyarrow
+    **dict.fromkeys([pa.date32(), pa.date64()], "date"),
```
Comment (bubriks) [inline python/hsfs/engine/python.py:107]: `pa.date64()` not in tests (only `date32`)

**Review by bubriks — CHANGES_REQUESTED** [review]

**Review by bubriks — APPROVED** [review]

File: python/hsfs/engine/python.py [hunk 11]
```diff
@@ -82,6 +83,31 @@
...
+# Decimal types are currently not supported
+PYARROW_HOPSWORKS_DTYPE_MAPPING = {
+    **dict.fromkeys(
+        [pa.uint8(), pa.uint16(), pa.int8(), pa.int16(), pa.int32()], "int"
+    ),
+    **dict.fromkeys([pa.uint32(), pa.int64()], "bigint"),
+    **dict.fromkeys([pa.float16(), pa.float32()], "float"),
+    **dict.fromkeys([pa.float64()], "double"),
+    **dict.fromkeys(
+        [pa.timestamp("ns")] + [pa.timestamp("ns", tz=tz) for tz in pytz.all_timezones],
```
Comment (bubriks) [inline python/hsfs/engine/python.py:95]: this will have to be updated after this is merged: https://github.com/logicalclocks/feature-store-api/pull/1169/files

**Review by bubriks — CHANGES_REQUESTED** [review]
this will have to be updated after this is merged: https://github.com/logicalclocks/feature-store-api/pull/1169/files

**Review by bubriks — APPROVED** [review]
Looks good
